### PR TITLE
Understanding Hill Climbing Algorithm in Artificial Intelligence

### LINKED ISSUE #863: Understanding Hill Climbing Algorithm in Artificial Intelligence
### Brief Summary: 
Hill climbing algorithm is an AI algorithm that increases value continuously until a peak solution is achieved. This algorithm is utilized in the optimization of mathematical problems. This article provides an understanding of this important concept. 

### Key Takeaways: 
1. Introduction to hill climbing algorithm
2. Features of hill climbing algorithm
3. State space diagram analysis
4. Problems in hill climbing algorithm
5. Types of hill climbing algorithm
6. Conclusion

### References: 
Please list links to any published content/research that you intend to use to support/guide this article. (If none, please indicate 'N/A'.)

### PR DESCRIPTION
**Pull request by mbaabuones**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process and style-once published, you can then submit any number of articles.)**
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

This fixes #863

### CODE REVIEW

**Comment by lalith1403** [comment]
@mbaabuones Note the modifications made. I have converted passive voice into active voice. I observed many passive voice sentences, hence use these changes as a guide for future articles.

**Comment by mbaabuones** [comment]
> 
> 
> @mbaabuones Note the modifications made. I have converted passive voice into active voice. I observed many passive voice sentences, hence use these changes as a guide for future articles.

@lalith1403 Okay. Noted. Thank you.

File: articles/understanding-hill-climbing-in-ai/index.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+* **Flat local maximum:** This is a flat region where the neighboring solutions attain the same value. 
+* **Shoulder:** This is a plateau whose edge is stretching upwards. 
+
+### Problems in Hill Climbing 
+There are three regions in which a hill-climbing algorithm cannot attain a global maximum or the optimal solution: local maximum, ridge, and plateau. 
+
+#### Local maximum
+At this point, the neighboring states have lower values than the current state. The greedy approach feature will not move the algorithm to a worse off state. This will lead to the hill-climbing process's termination, even though this is not the best possible solution.
+
+This problem can be solved by searching for new paths to move from an undesirable state. This can be done through [backpropagation](https://en.wikipedia.org/wiki/Backpropagation).
```
Comment (lalith1403) [inline articles/understanding-hill-climbing-in-ai/index.md:43]: Computing BPA is done via gradient descent usually. There exist other methods to find the optimum cost function as well. However, SGD also suffers from local maxima and minima. Can you check this statement once? I believe BPA also suffers from the local maximum. For example, momentum, Adaptive Gradients, and Adam are optimizers modified to combat this problem. Put across your views in a detailed manner.

File: articles/understanding-hill-climbing-in-ai/index.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+![Ridge in Hill Climbing](/engineering-education/understanding-hill-climbing-in-ai/ridge-in-hill-climbing.png)
+
+[Image Source: VietMX Blog](https://www.maixuanviet.com/wp-content/uploads/2020/05/hill-climbing-algorithm-in-ai4.png)
+
+### Types of Hill Climbing Algorithm
+The following are the types of a hill-climbing algorithm:
+#### Simple Hill Climbing
+This is a simple form of hill climbing that evaluates the neighboring solutions. If the next neighbor state has a higher value than the current state, the algorithm will move. The neighboring state will then be set as the current one. This algorithm consumes less time and requires little computational power. However, the solutions produced by the algorithm are sub-optimal. In some cases, an optimal solution may not be guaranteed.  
+
+##### Algorithm
```
Comment (lalith1403) [inline articles/understanding-hill-climbing-in-ai/index.md:68]: Include a GIF if possible for all types of hill-climbing algorithms.

**Review by lalith1403 — CHANGES_REQUESTED** [review]

**Comment by hectorkambow** [comment]
plagiarism check done

File: articles/understanding-hill-climbing-in-ai/index.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+* **No Backtracking:** A hill-climbing algorithm only works on the current state and succeeding states (future). It does not look at the previous states. 
+* **Feedback mechanism:** The algorithm has a feedback mechanism that helps it decide on the direction of movement (whether up or down the hill). The feedback mechanism is enhanced through the [generate-and-test technique](http://intelligence.worldofcomputing.net/ai-search/generate-and-test-search.html#:~:text=%20Algorithm%3A%20Generate-And-Test%20%201%201.Generate%20a%20possible,quit%20else%20go%20to%20step%201.%20More%20). 
+* **Incremental change:** The algorithm improves the current solution by incremental changes. 
+ 
+### State-Space Diagram Analysis
+A state-space diagram provides a graphical representation of states and the cost or objective functions. If the objective function is the y-axis, we aim to establish the local maximum and global maximum. If the cost function represents this axis, we aim to establish the local minimum and global minimum. More information about local minimum, local maximum, global minimum, and global maximum can be found [here](https://en.wikipedia.org/wiki/Maxima_and_minima).
```
[the remark this answers is not in the document]
Comment (mbaabuones) [reply articles/understanding-hill-climbing-in-ai/index.md]: I have modified it.

File: articles/understanding-hill-climbing-in-ai/index.md (hunk 1 above)
Comment (mbaabuones) [reply articles/understanding-hill-climbing-in-ai/index.md]: You are right. I have revised this part.

File: articles/understanding-hill-climbing-in-ai/index.md (hunk 2 above)
Comment (mbaabuones) [reply articles/understanding-hill-climbing-in-ai/index.md:67]: I couldn't find a suitable GIF or image for all the types of hill climbing algorithms.

**Comment by mbaabuones** [comment]
@lalith1403 I finished working on the changes.

**Pull request merged**